File: spring-boot-iot-ui/src/api/ruleDefinition.ts
import { request } from './request';
import { buildQueryString } from './query';
import type { ApiEnvelope, IdType } from '../types/api';

export type RuleDefinitionScope = 'METRIC' | 'PRODUCT_TYPE' | 'PRODUCT' | 'DEVICE' | 'BINDING';
export type RuleDefinitionScopeView = 'BUSINESS' | 'SYSTEM' | 'ALL';

/**
 * 阈值规则配置 API
 */

// 阈值规则接口定义
export interface RuleDefinition {
      id: IdType;
      riskMetricId?: IdType | null;
      ruleScope?: RuleDefinitionScope | string | null;
      productType?: string | null;
      productId?: IdType | null;
      deviceId?: IdType | null;
      riskPointDeviceId?: IdType | null;
      ruleName: string;
      metricIdentifier: string;
      metricName: string;
      expression: string;
      duration: number;
      alarmLevel: string;
      notificationMethods: string;
      convertToEvent: number;
      status: number;
      tenantId: IdType;
      remark: string;
      createBy: number;
      createTime: string;
      updateBy: number;
      updateTime: string;
      deleted: number;
}

export interface RuleDefinitionPageResult {
      total: number;
      pageNum: number;
      pageSize: number;
      records: RuleDefinition[];
}

export interface RuleDefinitionBatchAddItem {
      index: number;
      ruleId?: IdType | null;
      ruleName?: string | null;
      metricIdentifier?: string | null;
      success?: boolean | null;
      message?: string | null;
}

export interface RuleDefinitionBatchAddResult {
      totalCount: number;
      successCount: number;
      failedCount: number;
      items: RuleDefinitionBatchAddItem[];
}

export interface RuleDefinitionEffectivePreviewCandidate {
      ruleId?: IdType | null;
      ruleName?: string | null;
      ruleScope?: RuleDefinitionScope | string | null;
      ruleScopeText?: string | null;
      scopeTarget?: string | null;
      metricIdentifier?: string | null;
      metricName?: string | null;
      expression?: string | null;
      alarmLevel?: string | null;
      status?: number | null;
      priority?: number | null;
      matchedContext?: boolean | null;
      selected?: boolean | null;
      reason?: string | null;
}

export interface RuleDefinitionEffectivePreview {
      tenantId?: IdType | null;
      riskMetricId?: IdType | null;
      metricIdentifier?: string | null;
      productId?: IdType | null;
      productType?: string | null;
      deviceId?: IdType | null;
      riskPointDeviceId?: IdType | null;
      hasMatchedRule?: boolean | null;
      matchedScope?: RuleDefinitionScope | string | null;
      matchedScopeText?: string | null;
      decision?: string | null;
      matchedRule?: RuleDefinition | null;
      candidates?: RuleDefinitionEffectivePreviewCandidate[];
}

// 获取规则列表
export const getRuleList = (params?: {
      riskMetricId?: IdType;
      ruleName?: string;
      metricIdentifier?: string;
      alarmLevel?: string;
      status?: number;
      ruleScope?: RuleDefinitionScope | string;
      scopeView?: RuleDefinitionScopeView | string;
      productType?: string;
      productId?: IdType;
}): Promise<ApiEnvelope<RuleDefinition[]>> => {
      const queryString = buildQueryString(params);
      const path = queryString ? `/api/rule-definition/list?${queryString}` : '/api/rule-definition/list';
      return request<RuleDefinition[]>(path, { method: 'GET' });
};

// 分页获取规则列表
export const pageRuleList = (params?: {
      riskMetricId?: IdType;
      ruleName?: string;
      metricIdentifier?: string;
      alarmLevel?: string;
      status?: number;
      ruleScope?: RuleDefinitionScope | string;
      scopeView?: RuleDefinitionScopeView | string;
      productType?: string;
      productId?: IdType;
      pageNum?: number;
      pageSize?: number;
}): Promise<ApiEnvelope<RuleDefinitionPageResult>> => {
      const queryString = buildQueryString(params);
      const path = queryString ? `/api/rule-definition/page?${queryString}` : '/api/rule-definition/page';
      return request<RuleDefinitionPageResult>(path, { method: 'GET' });
};

// 获取规则详情
export const getRuleById = (id: IdType): Promise<ApiEnvelope<RuleDefinition>> => {
      return request<RuleDefinition>(`/api/rule-definition/get/${id}`, { method: 'GET' });
};

export const previewEffectiveRule = (params: {
      tenantId?: IdType;
      riskMetricId?: IdType;
      metricIdentifier?: string;
      productId?: IdType;
      productType?: string;
      deviceId?: IdType;
      riskPointDeviceId?: IdType;
}): Promise<ApiEnvelope<RuleDefinitionEffectivePreview>> => {
      const queryString = buildQueryString(params);
      const path = queryString
            ? `/api/rule-definition/effective-preview?${queryString}`
            : '/api/rule-definition/effective-preview';
      return request<RuleDefinitionEffectivePreview>(path, { method: 'GET' });
};

// 新增规则
export const addRule = (data: Partial<RuleDefinition>): Promise<ApiEnvelope<RuleDefinition>> => {
      return request<RuleDefinition>('/api/rule-definition/add', { method: 'POST', body: data });
};

// 更新规则
export const addRuleBatch = (data: Array<Partial<RuleDefinition>>): Promise<ApiEnvelope<RuleDefinitionBatchAddResult>> => {
      return request<RuleDefinitionBatchAddResult>('/api/rule-definition/batch-add', { method: 'POST', body: data });
};

export const updateRule = (data: Partial<RuleDefinition>): Promise<ApiEnvelope<RuleDefinition>> => {
      return request<RuleDefinition>('/api/rule-definition/update', { method: 'POST', body: data });
};

// 删除规则
export const deleteRule = (id: IdType): Promise<ApiEnvelope<void>> => {
      return request<void>(`/api/rule-definition/delete/${id}`, { method: 'POST' });
};

File: spring-boot-iot-alarm/src/main/java/com/ghlzm/iot/alarm/entity/RuleDefinition.java
package com.ghlzm.iot.alarm.entity;

import com.baomidou.mybatisplus.annotation.IdType;
import com.baomidou.mybatisplus.annotation.TableId;
import com.baomidou.mybatisplus.annotation.TableName;
import lombok.Data;

import java.io.Serial;
import java.io.Serializable;
import java.time.LocalDateTime;

/**
 * 阈值规则配置实体
 */
@Data
@TableName("rule_definition")
public class RuleDefinition implements Serializable {
      @Serial
      private static final long serialVersionUID = 1L;

      /**
       * 主键ID
       */
      @TableId(type = IdType.AUTO)
      private Long id;

      /**
       * 规则名称
       */
      private String ruleName;

      /**
       * 测点标识符
       */
      private Long riskMetricId;

      /**
       * 测点标识符
       */
      private String metricIdentifier;

      /**
       * 测点名称
       */
      private String metricName;

      /**
       * 表达式（如：value > 100）
       */
      private String expression;

      /**
       * 持续时间（秒）
       */
      private Integer duration;

      /**
       * 告警等级：critical-严重, warning-警告, info-提醒
       */
      private String alarmLevel;

      /**
       * 通知方式：email,sms,wechat
       */
      private String notificationMethods;

      /**
       * 是否转事件：0-否，1-是
       */
      private Integer convertToEvent;

      /**
       * 状态：0-启用，1-停用
       */
      private Integer status;

      /**
       * 租户ID
       */
      private Long tenantId;

      /**
       * 创建时间
       */
      private LocalDateTime createTime;

      /**
       * 更新时间
       */
      private LocalDateTime updateTime;

      /**
       * 创建人
       */
      private Long createBy;

      /**
       * 更新人
       */
      private Long updateBy;

      /**
       * 逻辑删除：0-未删除，1-已删除
       */
      private Integer deleted;
}
